crates/sp-executor/src/lib.rs

```diff
@@ -102,6 +113,62 @@ impl<Hash: Encode> BundleHeader<Hash> {
     }
 }
 
+// TODO: unify the type between the solution and threshold better.
+pub fn derive_bundle_election_solution(domain_id: DomainId, vrf_output: &[u8]) -> u64 {
+    let mut data = domain_id.to_le_bytes().to_vec();
+    data.extend_from_slice(vrf_output);
+
+    let local_domain_randomness = blake2b_256_hash(&data);
+
+    let election_solution = u64::from_le_bytes(
+        local_domain_randomness
+            .split_at(core::mem::size_of::<u64>())
+            .0
+            .try_into()
+            .expect("Local domain randomness must fit into u64; qed"),
+    );
+
+    election_solution
+}
+
+pub fn calculate_bundle_election_threshold(
+    stake_weight: StakeWeight,
+    total_stake_weight: StakeWeight,
+    slot_probability: (u64, u64),
+) -> u128 {
+    stake_weight / total_stake_weight * u128::MAX * u128::from(slot_probability.0)
+        / u128::from(slot_probability.1)
```

Comment:
This will always be zero assuming `stake_weight` is never as big as `total_stake_weight`. Also `* u128::MAX` will cause overflow unless `stake_weight / total_stake_weight` is `0` or `1`.

This is problematic, I think you need to rewrite it correctly, possibly like this:
```suggestion
    u128::MAX / u128::from(slot_probability.1) * u128::from(slot_probability.0) / total_stake_weight * stake_weight
```

And provide a comment what it is in more readable form, I did that in `pallet-subspace` already.